velox/docs/develop/anti-join.rst

```diff
@@ -0,0 +1,581 @@
+in the presence of NULLs in the outer query or subquery. NOT IN semantics are
+implemented by the null aware anti join. NOT EXISTS semantics are implemented
+by the regular anti join.
+
+Currently, Velox provides only null aware anti join via the JoinType::kAnti.
+Regular anti join is not available. We will rename kAnti to kNullAwareAnti for
+clarity and introduce a new kAnti join type for the regular anti join.
+
+This article explains the differences in semantics between NOT IN and NOT EXISTS
+queries and discusses the implementation of these in the null aware and regular
+anti joins.
+
+NOTE: Presto optimizer doesn’t plan NOT IN and NOT EXISTS queries using anti
+joins. To compensate, query plan translation in Prestissimo detects a NOT IN
+pattern and `converts <https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp#L1031>`_
+it into Velox anti join plan node.
+
+NOTE: `Substrait <https://substrait.io/relations/logical_relations/#join-types>`_
+currently defines only one type of anti join and doesn’t specify
+whether the semantics are NOT IN or NOT EXISTS. This topic is in active
+`discussion <https://github.com/substrait-io/substrait/issues/325>`_.
+
+It is easiest to understand the semantic differences using examples.
+
+Consider table t:
+
+==== =====
+id   value
+==== =====
+NULL 0
+1    1
+2    2
+==== =====
+
+and table u:
+
+==== =====
+id   value
+==== =====
+NULL 0
+2    2
+3    3
+==== =====
+
+
+When experimenting with queries, it is convenient to use WITH clause combined
+with UNNEST clause to create temporary datasets. Here is how to create and use
+the datasets above:
+
+.. code-block:: sql
+
+	WITH
+	     t as (SELECT * FROM unnest(array[null, 1, 2], array[0, 1, 2]) as _t(id, value)),
+	     u as (SELECT * FROM unnest(array[null, 2, 3], array[0, 2, 3]) as _t(id, value))
+	<query that uses t and u>
+
+	> SELECT * FROM unnest(array[null, 1, 2], array[0, 1, 2]) as _t(id, value);
+
+	  id  | value
+	------+-------
+	 NULL |     0
+	    1 |     1
+	    2 |     2
+	(3 rows)
+
+	> SELECT * FROM unnest(array[null, 2, 3], array[0, 2, 3]) as _t(id, value);
+
+	  id  | value
+	------+-------
+	 NULL |     0
+	    2 |     2
+	    3 |     3
+	(3 rows)
+
+	> WITH
+		t as (SELECT * FROM unnest(array[null, 1, 2], array[0, 1, 2]) as _t(id, value)),
+		u as (SELECT * FROM unnest(array[null, 2, 3], array[0, 1, 2]) as _t(id, value))
+	SELECT * FROM t FULL JOIN u ON t.id = u.id;
+
+	  id  | value |  id  | value
+	------+-------+------+-------
+	    1 |     1 | NULL | NULL
+	    2 |     2 |    2 |     1
+	 NULL | NULL  |    3 |     2
+	 NULL |     0 | NULL | NULL
+	 NULL | NULL  | NULL |     0
+	(5 rows)
+
+
+NOT IN <subquery> Semantics
+---------------------------
+
+We need to consider 3 cases of NOT IN queries:
+
+* Subquery returns a row with a NULL.
+* Subquery returns no rows.
+* Subquery returns one or more rows and no row has a NULL.
```

Comment:
I believe this would be the same as the first case. No?

Reply:
I had read the first case was exclusively just a single NULL, since the 3rd bullet point said one or more rows and no row has a NULL. The rest of the explanation made that clear though.